### DISCUSSION TITLE
Some feedback & possible evolutions ideas by @jnastarot

### DISCUSSION THREAD

**Discussion by pthom** [pthom/litgen#28] [Ideas]
Below, a copy of a message shared by @jnastarot in [this issue](https://github.com/pthom/litgen/pull/27#issuecomment-2512879710)

---

Hello,
Thank you so much for this awesome project! I truly appreciate the effort that has gone into developing Litgen.

I would like to share my experience and provide some suggestions
# Litgen is Useful

Litgen has been much more useful to me compared to tools like SWIG and Rosetta's Binder. The main reason is its flexibility in configuring the processing flow, which allows for more tailored adjustments.
I tried SWIG and Binder, but the bindings they generate are difficult to fix manually

# Thoughts

However, I feel that litgen is still in its early stages and could benefit from additional features and improvements.

* Replace regex configuration with callbacks
    Using callbacks instead of regular expressions would enable more robust decision-making. Callbacks allow developers to perform additional checks and apply specific fixes dynamically, which is not always feasible with regex-based configurations.

* Add more options
    Features like those proposed in https://github.com/pthom/litgen/pull/27 are essential. Developers using ltgen often have diverse use cases, and more options would make the tool adaptable to their specific needs.

* Minimize header amalgamation
    Header amalgamation is not always desirable, especially in cases involving class derivation. Reducing this dependency or making it optional would enhance flexibility.

* My use case
    I am working on a pet project involving file format dissection. The project includes many files containing numerous classes organized under different namespaces. Header files have specific prefixes, and there are various definitions, such as struct and enum. Sometimes, structures are defined using typedef, and I believe this case should be handled natively by litgen https://github.com/pthom/litgen/issues/7.

* Since my project has a large number of classes, files, and namespaces, I expect bindings to be separated into submodules. Currently, I achieve this by post-processing generated files, using markers like <generated_from> and <submodule>. It would be fantastic if litgen could handle submodule separation out of the box.

* Performance
    I have noticed that litgen slows down significantly when processing many enums. For example, processing all my headers (~2.5 MB) takes around 40 minutes. I attempted to profile the issue but couldn’t quickly identify the bottleneck. This might be worth investigating further.

I believe litgen has the potential to become as popular and essential as tools like Pybind11 or Nanobind, as well as a valuable complement to other binding tools
Once again, thank you! Litgen has saved me a lot of time

**Comment by jnastarot** [comment]
I'm also thinking about including a custom `__new__` to handle classes with `std::enable_shared_from_this`. (See https://nanobind.readthedocs.io/en/latest/classes.html#custom-new.)  
Maybe it's a bit much, but I think a lot of people use `std::shared_ptr` and expect objects created in Python to be managed as `shared_ptr` so they can use them in C++.  

I know `srcml` might not handle inheritance in C++ properly, and `std::enable_shared_from_this` might not be supported. So, I’m thinking about adding a callback for this feature, like it was done in https://github.com/pthom/litgen/pull/27.

**Reply by pthom** [reply]
This is quite an advanced use case, but interesting. Feel free to explore it and ping me if you have interesting results (with tests of course :-)

**Comment by lczech** [comment]
Hi @pthom,

Thank you for creating litgen! Just the other day, I discovered it, and started to have a look at it, as a potential alternative to Rosetta's Binder. It does seem to solve some current shortcomings of Binder, and allow for more flexibility, which got me curious. However, for my use case, it also seems to have some (unfortunately significant) downsides, and I hence would like to add to the above discussion with my thoughts and suggestions. Let me know if either of those are worthy of opening an issue about them as well.

In short, I am trying to create bindings for my (rather large) bioinformatics library [genesis](https://github.com/lczech/genesis), which has around 100k lines of code by now. If I get this to work, it would instantly double your current [battle test size](https://github.com/pthom/litgen?tab=readme-ov-file#battle-tested) with a single new library :-)

Here are my findings and thoughts while trying to get started with litgen:


## Setup and getting started

  - I found it a bit hard to actually "get started". The [quick start](https://pthom.github.io/litgen/litgen_book/litgen_template/README.html) starts amazingly with the super nice and short snipped for final users. However, it then goes into quite some detail that is necessary to get the template repo to work. What I was instead hoping for was a simple script that takes a directory path as input, finds all cpp headers, and generates bindings for them. That script could be the single entry point for creating bindings, so that I as a developer would only need to adjust that script, instead of having to understand the complex structure of the template (which I think I do not fully get yet).

  - The template repo itself is a bit confusing as well. It is a submodule of the main litgen repo, but then it pulls litgen again as its own dependency when going through the [install steps](https://pthom.github.io/litgen/litgen_book/litgen_template/README.html#install-requirements)? I'd suggest to simplify that :-)

  - The `autogenerate_bindings.py` script in the template is going in the direction of the above idea of a single script, and I suggest to make it the main script in the main litgen repo instead, which I as a developer would then adjust to my needs. 

  - However, the script would probably need some adjustments as well: It calls a function `write_generated_code_for_files` that does not seem to be officially documented in the "book". Searching for that function shows that it pops up on [some not-listed page](https://pthom.github.io/litgen/litgen_book/example_bindings/DasLib/Readme.html#generate-the-bindings-for-a-collection-of-files), which I assume is some old content? So, I am not sure about this.


## Usability and major suggestions

  - Adding to the suggestion of @jnastarot about the header amalgation: I also would like to suggest to allow for more flexibility there. In particular in combination with the comment on performance above, I am a bit concerned about the runtime of litgen: My codebase is ~7.3 MB, so extrapolating from @jnastarot, I'm looking at ~2h runtime for litgen, which is not encouraging. On top of that, compiling one huge file might also be prohibitive in terms of memory, and not good for incremental builds.
  
    My suggestion is hence to add a mechanism to parse each header separately, and write out bindings for each to an individual compilation unit, with a central `main` file that defines the pybind/nanobind module, and includes calls to functions defined in those compilation units to add bindings for whatever stuff their corresponding original headers defined. This would be similar to how Binder deals with that, and also just be a bit cleaner to work with, if I have a 1-to-1 correspondence between my ~400 headers and the auto-generated bindings code, instead of one humongous file containing all the bindings.

    Ideally, this would be set it up such that incremental builds also work: If the generated bindings file for a header has a newer timestamp then the header it is based on, it is not updated, so that the timestamp does not change, and an incremental compilation does not re-compile the unit. I have set this up in CMake for Binder already, and it allows me to incrementally work on my bindings, instead of having to re-compile everything every time. For litgen, this would also mean that only changed code needs to be processed, which should eliminate the issue above about performance.

    The slightly tricky part here of course are then class hierarchies where the classes are in separate files (as I think is commonly the case). But maybe there is a way around that as well?!

  - My second major suggestion are custom bindings: I don't think that litgen (or any automated tool for that matter) will be able to auto-generate all special cases, and so I need a way to define additional bindings manually. This means, adding custom binding code for classes and functions to a namespace module, as well as adding additional methods as class member functions to a class that is otherwise receiving auto-generated bindings.

    I am not sure how complicated this is to add to litgen, but I could imagine that a similar logic to Binder could work, with `addons` being definable that add functions to be called, taking the pybind/nanobind module or class instance as arguments, and allowing to add custom bindings to it.

## Confusions and minor suggestions

  - The into explains how [litgen only works for C++14](https://pthom.github.io/litgen/litgen_book/00_00_intro.html#compatibility), but then later mentions [bindings for threeway comparisons](https://pthom.github.io/litgen/litgen_book/06_05_00_classes.html?highlight=spaceship#overloading-comparisons-with-the-spaceship-operator), which is a [C++20 feature](https://en.cppreference.com/w/cpp/language/operator_comparison.html). What does that imply then?

  - Similarly, I'd be interested in being able to bind `std::string_view` and other additions that require later versions of the standard. Is that something that would need to be added in srcML first, and could then be added to litgen later?

  - On that note, I am getting several parsing errors from srcML on some complicated C++ constructs in my code, which just cause the whole process to terminate before producing the bindings. Maybe those exceptions could be handled more gracefully, i.e., get caught and warned about, instead of terminating.  Apart from that, I guess that I should report such parsing issues to the [srcML repo](https://github.com/srcML/srcML/issues) directly?

  - My codebase uses [Doxygen](https://doxygen.nl/) for documentation, which however does not seem to be properly supported by litgen - all comments are just passed on to the binding documentation as they are, instead of, e.g., stripping the `@` annotation tags of Doxygen. Just a minor thing, but might be relevant for others as well.

  - Adding to the suggestion of @jnastarot above, I also find the regex configuration a bit unintuitive to use. I was wondering what the best way is for my case, where I will need to specify the return value policy for probably hundreds of functions. I'm currently thinking to use it like that:

    ```
    options.fn_return_force_policy_reference_for_pointers__regex += r"^my_function$"
    options.fn_return_force_policy_reference_for_pointers__regex += r"^another_function$"
    ```

    That is, keep adding function names for each one that I need to be listed there. If that's the case, then a simple list as an additional way of specifying functions and classes for special treatment might also work:

    ```
    options.fn_return_force_policy_reference_for_pointers_list += [ 'my_function' ]
    options.fn_return_force_policy_reference_for_pointers_list += [ 'another_function' ]
    ```

    This could easily be iterated over and added to the regex internally in litgen after having parsed the options, so that internal code would not even have to change.

  - While at it, what is the double underscore before, e.g., `__regex` about in `fn_return_force_policy_reference_for_pointers__regex`? I am not a power user of Python, but it seems weird to me :-)


That's a lot - I hope that's okay! If you think that my major points can be addressed, I could see if I find the time to support the development here a bit as well, as I'd like to get those bindings to work eventually, and it seems easier than starting my own project for auto-generating them :-)

Thanks a lot in advance, and looking forward to hearing your thoughts
Lucas

**Comment by pthom** [comment]
Hi @lczech 

Thanks for your feedback. Your message is quite long message with many questions and remarks. 
 As I am on holiday, I will need a few days before answering.

**Reply by lczech** [reply]
Hi @pthom, no worries, and no rush, just wanted to give a bit of feedback that I hope might be useful.

**Comment by pthom** [comment]
Hi @lczech 

Thanks for your feedback. Your message is quite long message with many questions and remarks. 
 As I am on holiday, I did not have time to answer all with details.

I will answer in several parts, since your message was quite long. See below.

**Comment by pthom** [comment]
@lczech : first part of my answers, on the setup part

## Setup and getting started
> 
> What I was instead hoping for was a simple script that takes a directory path as input, finds all cpp headers, and generates bindings for them. That script could be the single entry point for creating bindings, so that I as a developer would only need to adjust that script, instead of having to understand the complex structure of the template (which I think I do not fully get yet).

litgen goes beyond creating the C++ binding code as it also creates the stub files and C++/Python glue code to make the bindings more usable. This goes far beyond the scope of a single script.

The template provides guidance on best practices to provide bindings for a large library, including guidance for CMake and scikit-build, how to build wheels for different platforms (using cibuildwheel), etc. 

However, if you feel inclined to create a simpler template, I would be happy to review a PR.

>* The template repo itself is a bit confusing as well. It is a submodule of the main litgen repo, but then it pulls litgen again as its own dependency when going through the [install steps](https://pthom.github.io/litgen/litgen_book/litgen_template/README.html#install-requirements)? I'd suggest to simplify that :-)

The template repo is a submodule of litgen for my own comfort, so that I do not forget to update it when needed. I hear your suggestion, but IMO the need is not urgent. 

> * The `autogenerate_bindings.py` script in the template is going in the direction of the above idea of a single script, and I suggest to make it the main script in the main litgen repo instead, which I as a developer would then adjust to my needs.

My opinion is that this script should be shown more prominently in the doc. 
I updated the doc to mention it in the "Quickstart with litgen" section.
See: https://pthom.github.io/litgen/litgen_book/litgen_template/README.html#customize-bindings-generation

**Comment by pthom** [comment]
@lczech : follow-up to the answers

## Confusions and minor suggestions

### C++14 or later?
> * The into explains how [litgen only works for C++14](https://pthom.github.io/litgen/litgen_book/00_00_intro.html#compatibility), but then later mentions [bindings for threeway comparisons](https://pthom.github.io/litgen/litgen_book/06_05_00_classes.html?highlight=spaceship#overloading-comparisons-with-the-spaceship-operator), which is a [C++20 feature](https://en.cppreference.com/w/cpp/language/operator_comparison.html). What does that imply then?

Simply that I added support for it as a convenience. srcML is still limited to C++14, but I could parse the spaceship operator as a special case.

### Binding more modern C++ features
> * Similarly, I'd be interested in being able to bind `std::string_view` and other additions that require later versions of the standard. Is that something that would need to be added in srcML first, and could then be added to litgen later?

I do think that a function using `string_view` as a param will be supported by litgen and nanobind. Feel free to try it out.
 
> * On that note, I am getting several parsing errors from srcML on some complicated C++ constructs in my code, which just cause the whole process to terminate before producing the bindings. Maybe those exceptions could be handled more gracefully, i.e., get caught and warned about, instead of terminating.  

Exceptions should be handled, it is normally the case. Feel free to post a minimal example that causes a crash, and it could be studied.

> Apart from that, I guess that I should report such parsing issues to the [srcML repo](https://github.com/srcML/srcML/issues) directly?

Yes, please do so. Be aware that srcML authors do not always answer quickly, so it may take a while before your issue is addressed.


> * My codebase uses [Doxygen](https://doxygen.nl/) for documentation, which however does not seem to be properly supported by litgen - all comments are just passed on to the binding documentation as they are, instead of, e.g., stripping the `@` annotation tags of Doxygen. Just a minor thing, but might be relevant for others as well.

Please post a minimal example that could be studied.

**Comment by pthom** [comment]
@lczech : follow-up to the answers

## Splitting up into multiple compilation units

### Speculative estimation of litgen runtime
>* Adding to the suggestion of @jnastarot about the header amalgation: I also would like to have more flexibility there! In particular in combination with the comment on performance above, I am a bit concerned about the runtime of litgen: My codebase is ~7.3 MB, so extrapolating from @jnastarot, I'm looking at ~2h runtime for litgen, which is not encouraging. On top of that, compiling one huge file might also be prohibitive in terms of memory, and not good for incremental builds.

Did you try it on your side before giving this 2h estimation?!? On my side, I can generate bindings for [ImGui](https://github.com/ocornut/imgui/blob/master/imgui.h) (a huge header file of 450KB) in less than ten seconds...

 @jnastarot was mentionning issues with enums, not all C++ code. Furthermore, litgen will only process header files, not C++ source files.

### Splitting into multiple compilation units

>  My suggestion is hence to add a mechanism to parse each header separately, and write out bindings for each to an individual compilation unit, with a central `main` file that defines the pybind/nanobind module, and includes calls to functions defined in those compilation units to add bindings for whatever stuff their corresponding original headers defined. This would be similar to how Binder deals with that, and also just be a bit cleaner to work with, if I have a 1-to-1 correspondence between my ~400 headers and the auto-generated bindings code, instead of one humongous file containing all the bindings.

You may call litgen file per file, and this would do the trick. I do it for ImGui Bundle on my side: ImGui Bundle comprises 20 different libraries, for which the generation of bindings is done per library.

I would recommend doing it on a per "directory / sublibrary" basis, so that you can group related headers together, not on a per-header basis: handling 400 compilation units is not a good idea.

>  The slightly tricky part here of course are then class hierarchies where the classes are in separate files (as I think is commonly the case). But maybe there is a way around that as well?!

I recommend using an amalgamated header file for each "directory / sublibrary", and to use litgen tooling to generate the amalgamated header and its bindings.

If your intent is to generate bindings separately for each 400 header files, litgen will not be able to link classes defined in different headers: a file needs to be parsed for litgen to know about the classes it defines! (this is not new, and quite related to the way #include works in C/C++).

**Comment by pthom** [comment]
@lczech: follow up to the answers

### Custom bindings
> * My second major requirment are custom bindings: I don't think that litgen (or any automated tool for that matter) will be able to auto-generate all special cases, and so I need a way to define additional bindings manually. This means, adding custom binding code for classes and functions to a namespace module, as well as adding additional methods as class member functions to a class that is otherwise receiving auto-generated bindings.
>       I am not sure how complicated this is to add to litgen, but I could imagine that a similar logic to Binder could work, with `addons` being definable that add functions to be called, taking the pybind/nanobind module or class instance as arguments, and allowing to add custom bindings to it.
>

I do agree this is a part where a mechanism to generate custom code would be appreciable.

**Comment by pthom** [comment]
@lczech / end of my answers


## Configuration via regex / list / callbacks

> I also find the regex configuration a bit unintuitive to use. I was wondering what the best way is for my case, where I will need to specify the return value policy for probably hundreds of functions. I'm currently thinking to use it like that:
>   ```
>   options.fn_return_force_policy_reference_for_pointers__regex += r"^my_function$"
>   options.fn_return_force_policy_reference_for_pointers__regex += r"^another_function$"
>   ```

`codemanip` (installed together with litgen) provides some utilities to make regex construction easier.

For example, you can use `code_utils.join_string_by_pipe_char` to join multiple regex patterns with the pipe character `|` (which means "or" in regex). You can also use `code_utils.make_regex_var_name_contains_word` to create a regex pattern that matches variable names containing a specific word. Look at code_utils.py in the codemanip package for more regex utilities.

Example:
```python
from codemanip import code_utils

my__regex: str = code_utils.join_string_by_pipe_char(
    [
        code_utils.make_regex_var_name_contains_word("nb"),
        code_utils.make_regex_var_name_contains_word("size"),
        code_utils.make_regex_var_name_contains_word("count"),
        code_utils.make_regex_var_name_contains_word("total"),
        code_utils.make_regex_var_name_contains_word("n"),
    ]
)
```

>  what is the double underscore before, e.g., __regex

That it is a regex variable. The double underscore is just a naming convention to indicate that the variable holds a regex pattern.

I know that the configuration via regex is not the most user-friendly. Sometimes, I wonder if all those regex configurations 

```python
class LitgenOptions:
    ...
    fn_exclude_by_name__regex: str = ""
```

could be replaced by a more user-friendly approach, like allowing the user to provide a callable (function) that takes a string (the function name) and returns a boolean indicating whether to exclude it or not.:

```python
RegexOrStrMatcher = str | Callable[[str], bool]

class LitgenOptions:
    ...
    fn_exclude_by_name__regex: RegexOrStrMatcher = ""
```

I still need to think about this, since the __regex suffix would become confusing, but I also need to think about backward compatibility. This is not on my top priority list.

**Reply by pthom** [reply]
**Configuration via regex / list / callbacks**

@lczech : the configuration is now easier, since you can provide your own matcher under the form of a python function.
See [doc about options](https://pthom.github.io/litgen/litgen_book/02_03_00_options.html) and related commit: https://github.com/pthom/litgen/commit/42a535f975f910837ce93d823084853bba88411c

**Comment by pthom** [comment]
**Added support for custom bindings**

@jnastarot , @lczech : 

I just added a big feature which could help you a lot.

 You can now manually extend the generated bindings without modifying your C++ headers.  
  - `LitgenOptions.custom_bindings.add_custom_code_to_class(...)`  
  - `LitgenOptions.custom_bindings.add_custom_code_to_submodule(...)`  
  - `LitgenOptions.custom_bindings.add_custom_code_to_main_module(...)`  
  This allows you to inject extra methods, functions, or properties into classes, namespaces, or the main module.  

This feature is documented [in this page](https://pthom.github.io/litgen/litgen_book/08_10_00_custom_bindings.html).
I also released [v0.21.0](https://github.com/pthom/litgen/releases/tag/v0.21.0)

I would be very interested in your feedback.
